In [ ]:
from google.colab import drive
drive.mount('/content/drive')

MessageError: ignored

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import matplotlib as mpl
import seaborn as sns
import plotly.express as px
from sklearn.decomposition import PCA
%matplotlib inline
import plotly.graph_objects as go
from matplotlib.image import imread
from sklearn.datasets import (make_blobs,
                              make_circles,
                              make_moons)
from sklearn.cluster import KMeans, SpectralClustering
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import silhouette_samples, silhouette_score

%matplotlib inline
sns.set_context('notebook')
plt.style.use('fivethirtyeight')
from warnings import filterwarnings
filterwarnings('ignore')

#PCA and our business

Hello Boss person! Today, we will be taking a look at some peices of data important to our business and seeing if they are correlated using something called PCA, principle component analysis.  

In a nutshell, PCA is a tool used for dimensionsality reduction on large datasets to by transforming a large set of variables into a components by capturing the most important parts of the data and seeing correlation of our data. By projecting our data down onto our principal components, we can calculate the amount of variance that is explained by the components and its individual parts that our data contributes to determine the most important features.


Essentially, we reduce dimensions to get the most important parts of our data with components. Cool!

#Housing

In [ ]:
import pandas as pd
import plotly.express as px
from sklearn.decomposition import PCA
from sklearn.datasets import load_boston
from sklearn.datasets import fetch_california_housing

housing = fetch_california_housing()

boston = load_boston()
df = pd.DataFrame(housing.data, columns=housing.feature_names)
df = df.loc[:, ['AveBedrms', 'AveRooms', 'HouseAge', 'Population',
       'AveOccup', 'MedInc']]
df.head(5)

ImportError: ignored

Here, in our housing marketing branch, I am using PCA on different variables on housing in California including the median Income, average amount of rooms, average amount of bedrooms, average amount of occupants, etc... To see of there is any correlation and to determine the most important parts of our data

In [ ]:

scaler=StandardScaler()
df=scaler.fit_transform(df)
n_components = 3

pca = PCA(n_components=n_components)
components = pca.fit_transform(df)

total_var = pca.explained_variance_ratio_.sum() * 100

labels = {str(i): f"PC {i+1}" for i in range(n_components)}
labels['color'] = 'Median Price'

fig = px.scatter_matrix(
    components,
    dimensions=range(n_components),
    labels=labels,
    title=f'Total Explained Variance: {total_var:.2f}%',
)
fig.update_traces(diagonal_visible=False)
fig.show()


pca.explained_variance_ratio_

Running PCA on our data we find that with three components we can only explain about 71.21% of our variance, which is related to explaining our important data and any correlation.

Using plotly, I can use a scatter matrix to graph each of my components and see if there are any notable features or obvious correlation.



So, around 71% of the variance of my data is explained in three dimensionally reduced components. Is this favorable? Well, Not really.

In [ ]:
fig = px.scatter_3d(
    components, x=0, y=1, z=2,
    title=f'Total Explained Variance: {total_var:.2f}%',
    labels={'0': 'PC 1', '1': 'PC 2', '2': 'PC 3'}
)
fig.show()
pca.explained_variance_ratio_

It sounds nice knowing I can dimensionally reduce my data to three dimensions and explain 71% of my data, but that means the other 29% of my data is in higher dimensions that I cannot see or show. Plus, I cannot tell if i can distinguish the 29% as noisy data, meaning data that is not that important, or if it's actually important to me.

The array below the 3d graph shows each of my components and their decimal approximations on how much each of them contribute to explaing the total variance. Notice that the first two components only explain about half of our data. This is not favorable as capturing the most important parts of our data should not be in higher dimensions, i.e. higher than my component count of three   

#Wine

Hello once again Boss person. As part of our associative branch, we also run important data analysis on wine to determine the important features of wine to make the best possible drink.

In [ ]:
df3 = pd.read_csv("/content/drive/MyDrive/Christian Kolker/WineQT.csv")
df3.head()
df3


fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00             1.9      0.076   
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
3              11.2             0.280         0.56             1.9      0.075   
4               7.4             0.700         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
1138            6.3             0.510         0.13             2.3      0.076   
1139            6.8             0.620         0.08             1.9      0.068   
1140            6.2             0.600         0.08             2.0      0.090   
1141            5.9             0.550         0.10             2.2      0.062   
1142            5.9             0.645         0.12             2.0      0.075   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
1138                 29.0                  40.0  0.99574  3.42       0.75   
1139                 28.0                  38.0  0.99651  3.42       0.82   
1140                 32.0                  44.0  0.99490  3.45       0.58   
1141                 39.0                  51.0  0.99512  3.52       0.76   
1142                 32.0                  44.0  0.99547  3.57       0.71   

      alcohol  quality  
0         9.4        5  
1         9.8        5  
2         9.8        5  
3         9.8        6  
4         9.4        5  
...       ...      ...  
1138     11.0        6  
1139      9.5        6  
1140     10.5        5  
1141     11.2        6  
1142     10.2        5  

[1143 rows x 12 columns]

Once again, we will use PCA to reduce our number of dimensions and determine the most important parts of our data. Using PCA, we can determine which chemical components are the most important to wine. Neat!

In [ ]:

scaler=StandardScaler()
df=scaler.fit_transform(df3)
n_components = 2

pca = PCA(n_components=n_components)
components_2 = pca.fit_transform(df3)

total_var = pca.explained_variance_ratio_.sum() * 100

labels = {str(i): f"PC {i+1}" for i in range(n_components)}
labels['color'] = 'Median Price'

fig = px.scatter(
    components_2, x=0, y=1,
    title=f'Total Explained Variance: {total_var:.2f}%',
    labels={'0': 'PC 1', '1': 'PC 2', '2': 'PC 3'}
)
fig.show()


pca.explained_variance_ratio_


array([0.94668078, 0.0477564 ])

Standardizing our data and Running PCA, we find that the first two components explain over 99% of our data. This is exactly what we are looking for when doing PCA! This means that when we reduce our dimensions into 2 components that we can graph into the picture above, we capture nearly all of our important data.

The array tells us that the first principal component accounts for 94% of our data and the second principal component captures the other 5%.

So, here is an interesting fact about this data. Since our first two principal components capture nearly all of our important data, the rest of whatever is not captured is essentially noise and can be disgarded.

In [ ]:

pca = PCA(n_components=2)
X = df3
components_u = pca.fit_transform(X)
#print(pca.explained_variance_ratio_)
#print(pca.feature_names_in_)

pca2 = PCA(svd_solver='auto', whiten=True)
pca2.fit(X)

#print(pca.components_)
#print(sum(pca.components_[0][0:6]))
#print(sum(pca.components_[1][0:6]))
#print(sum(pca.components_[0][0:6]) + sum(pca.components_[1][0:6]))
#print(sum(pca.explained_variance_ratio_))

#Print the PC percentages
print("Principle Components")
print("PC1: " + str(round(pca.explained_variance_ratio_[0] * 100, 2)) + "%")
print("PC2: " + str(round(pca.explained_variance_ratio_[1] * 100, 2)) + "%\n")


squares1 = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]
i = 0
for x in pca.components_[0]:
  squares1[i] = x * x
  i += 1

squares2 = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]
i = 0
for x in pca.components_[1]:
  squares2[i] = x * x
  i += 1

#print(sum(squares1))
#print(squares1)

#Print the feature percentages
print("Feature percentages in PC1")
print("fixed acidity: " + str(round(squares1[0] * 100, 2)) + "%")
print("volatile acidity: " + str(round(squares1[1] * 100, 2)) + "%")
print("citric acid: " + str(round(squares1[2] * 100, 2)) + "%")
print("residual sugar: " + str(round(squares1[3] * 100, 2)) + "%")
print("chlorides: " + str(round(squares1[4] * 100, 2)) + "%")
print("free sulfur dioxide: " + str(round(squares1[5] * 100, 2)) + "%")
print("total sulfur dioxide: " + str(round(squares1[6] * 100, 2)) + "%")
print("density: " + str(round(squares1[7] * 100, 2)) + "%")
print("pH: " + str(round(squares1[8] * 100, 2)) + "%")
print("sulphates: " + str(round(squares1[9] * 100, 2)) + "%")
print("alcohol: " + str(round(squares1[10] * 100, 2)) + "%")
print("quality: " + str(round(squares1[11] * 100, 2)) + "%\n")

print("Feature percentages in PC2")
print("fixed acidity: " + str(round(squares2[0] * 100, 2)) + "%")
print("volatile acidity: " + str(round(squares2[1] * 100, 2)) + "%")
print("citric acid: " + str(round(squares2[2] * 100, 2)) + "%")
print("residual sugar: " + str(round(squares2[3] * 100, 2)) + "%")
print("chlorides: " + str(round(squares2[4] * 100, 2)) + "%")
print("free sulfur dioxide: " + str(round(squares2[5] * 100, 2)) + "%")
print("total sulfur dioxide: " + str(round(squares2[6] * 100, 2)) + "%")
print("density: " + str(round(squares1[7] * 100, 2)) + "%")
print("pH: " + str(round(squares1[8] * 100, 2)) + "%")
print("sulphates: " + str(round(squares1[9] * 100, 2)) + "%")
print("alcohol: " + str(round(squares1[10] * 100, 2)) + "%")
print("quality: " + str(round(squares1[11] * 100, 2)) + "%")

features = ['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar', 'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density','pH', 'sulphates', 'alcohol', 'quality']

X = df3[features]


loadings = pca.components_.T * np.sqrt(pca.explained_variance_)
#fig = px.scatter(components_5, x=0, y=1)
total_var = pca.explained_variance_ratio_.sum() * 100

labels = {str(i): f"PC {i+1}" for i in range(n_components)}
labels['color'] = 'Median Price'

fig = px.scatter(
    components_u, x=0, y=1,
    title=f'Total Explained Variance: {total_var:.2f}%',
    labels={'0': 'PC 1', '1': 'PC 2', '2': 'PC 3'}
)

for i, feature in enumerate(features):
    fig.add_shape(
        type='line',
        x0=0, y0=0,
        x1=loadings[i, 0],
        y1=loadings[i, 1]
    )
    fig.add_annotation(
        x=loadings[i, 0],
        y=loadings[i, 1],
        ax=0, ay=0,
        xanchor="center",
        yanchor="bottom",
        text=feature,
    )


#fig = px.scatter(components_u, x=0, y=1, color_continuous_scale=px.colors.sequential.Blackbody)
fig.show()


Principle Components
PC1: 94.67%
PC2: 4.78%

Feature percentages in PC1
fixed acidity: 0.0%
volatile acidity: 0.0%
citric acid: 0.0%
residual sugar: 0.01%
chlorides: 0.0%
free sulfur dioxide: 4.54%
total sulfur dioxide: 95.44%
density: 0.0%
pH: 0.0%
sulphates: 0.0%
alcohol: 0.0%
quality: 0.0%

Feature percentages in PC2
fixed acidity: 0.08%
volatile acidity: 0.0%
citric acid: 0.0%
residual sugar: 0.01%
chlorides: 0.0%
free sulfur dioxide: 95.33%
total sulfur dioxide: 4.54%
density: 0.0%
pH: 0.0%
sulphates: 0.0%
alcohol: 0.0%
quality: 0.0%


Credit to my good friend Spencer Moon for this wonderful way of explaining each of the components in terms of the columns of our dataframe they make up.

Here, we use our principal components to calculate the sum of the squares of each the components and find how our data contributes to the variation captured by each of our components.

On the graph, we see two vectors, also called loadings, that have the greatest magnititude and point in different directions. These are the important features of our data. These two vectors are like this because they help to explain the most amount of variance in our data according to our components and the linear combination of our original data variables.

The points closer to the total sulfur dioxide arrow are the wine samples that associate more with total sulfur dioxide, and the same goes for free sulfur dioxide

In terms of wine, the total amount of sulfur dioxode seems to be the most important factor taking up 95% of the explained variance in the first principal component, while free sulfur dioxide is second.

In [ ]:
n_pcs = pca.components_.shape[0]
print("number of principal components: ", n_pcs)

important_components = [np.abs(pca.components_[i]).argmax() for i in range(n_pcs)]
print("indicies of important features: ", important_components)

important_features = [features[important_components[i]] for i in range(n_pcs)]
print("most important features ", important_features, "\n")

number of principal components:  2
indicies of important features:  [6, 5]
most important features  ['total sulfur dioxide', 'free sulfur dioxide'] 



Credit to Justin Marotta for this block of code.

What this does block of code does is it determines the most important features of our componenets.

First, it finds the number of components and assigns it to n_pcs.

Then, it will find the indices of the most important features on each component. This is done by using a for loop in the range of our number of components, n_pcs, and finds the largest absolute value of each explained variance. Using list comprehension, these important components can be put into a list quite easily.

Finally, after having the indices, we can locate the columns of our data that are the most important features. This confirms our data and our graph with the loadings above.

#Breast Cancer data analysis

As another part of our other department branch, we analyze cancer cells to determine the important and distinct features of each of our measurements and data.

In [ ]:
df5 = pd.read_csv("/content/drive/MyDrive/Christian Kolker/data.csv")
df5 = df5.drop('id', 1)
df5 = df5.drop('diagnosis', 1)
df5 = df5.drop('Unnamed: 32', 1)

df5.columns


Index(['radius_mean', 'texture_mean', 'perimeter_mean', 'area_mean',
       'smoothness_mean', 'compactness_mean', 'concavity_mean',
       'concave points_mean', 'symmetry_mean', 'fractal_dimension_mean',
       'radius_se', 'texture_se', 'perimeter_se', 'area_se', 'smoothness_se',
       'compactness_se', 'concavity_se', 'concave points_se', 'symmetry_se',
       'fractal_dimension_se', 'radius_worst', 'texture_worst',
       'perimeter_worst', 'area_worst', 'smoothness_worst',
       'compactness_worst', 'concavity_worst', 'concave points_worst',
       'symmetry_worst', 'fractal_dimension_worst'],
      dtype='object')

Here, we have a whole list of data on breast cancer cells including the radius, perimeter, smoothness, compactness, concavitiy, etc..

In [ ]:
scaler=StandardScaler()
df=scaler.fit_transform(df5)
n_components = 2

pca = PCA(n_components=n_components)
components_5 = pca.fit_transform(df5)

total_var = pca.explained_variance_ratio_.sum() * 100

labels = {str(i): f"PC {i+1}" for i in range(n_components)}
labels['color'] = 'Median Price'

fig = px.scatter(
    components_5, x=0, y=1,
    title=f'Total Explained Variance: {total_var:.2f}%',
    labels={'0': 'PC 1', '1': 'PC 2', '2': 'PC 3'}
)
fig.show()
pca.explained_variance_ratio_

array([0.98204467, 0.01617649])

Upon executing PCA, we find that the first principal component captures 98% of the explained variance. WOW! This means 99% of our variance is within the first two components. So, the most important parts of our data is the first two dimensions which is what we reduced to. Therefore, nearly all the other higher dimensional data of the other components are just noise!

In [ ]:
pca = PCA(n_components=2)
X = df5.loc[:, ['compactness_se', 'concavity_se', 'concave points_se', 'symmetry_se',
       'fractal_dimension_se', 'radius_worst', 'texture_worst',]]
components_u = pca.fit_transform(X)
#print(pca.explained_variance_ratio_)
#print(pca.feature_names_in_)

pca2 = PCA(svd_solver='auto', whiten=True)
pca2.fit(X)

#print(pca.components_)
#print(sum(pca.components_[0][0:6]))
#print(sum(pca.components_[1][0:6]))
#print(sum(pca.components_[0][0:6]) + sum(pca.components_[1][0:6]))
#print(sum(pca.explained_variance_ratio_))

#Print the PC percentages
print("Principle Components")
print("PC1: " + str(round(pca.explained_variance_ratio_[0] * 100, 2)) + "%")
print("PC2: " + str(round(pca.explained_variance_ratio_[1] * 100, 2)) + "%\n")


squares1 = [1, 2, 3, 4, 5, 6, 7]
i = 0
for x in pca.components_[0]:
  squares1[i] = x * x
  i += 1

squares2 = [1, 2, 3, 4, 5, 6, 7]
i = 0
for x in pca.components_[1]:
  squares2[i] = x * x
  i += 1

#print(sum(squares1))
#print(squares1)

#Print the feature percentages
print("Feature percentages in PC1")
print("compactness_se: " + str(round(squares1[0] * 100, 2)) + "%")
print("concavity_se: " + str(round(squares1[1] * 100, 2)) + "%")
print("concave points_se: " + str(round(squares1[2] * 100, 2)) + "%")
print("symmetry_se: " + str(round(squares1[3] * 100, 2)) + "%")
print("fractal_dimension_se: " + str(round(squares1[4] * 100, 2)) + "%")
print("radius_worst: " + str(round(squares1[5] * 100, 2)) + "%")
print("texture_worst: " + str(round(squares1[6] * 100, 2)) + "%\n")

print("Feature percentages in PC2")
print("compactness_se: " + str(round(squares2[0] * 100, 2)) + "%")
print("concavity_se: " + str(round(squares2[1] * 100, 2)) + "%")
print("concave points_se: " + str(round(squares2[2] * 100, 2)) + "%")
print("symmetry_se: " + str(round(squares2[3] * 100, 2)) + "%")
print("fractal_dimension_se: " + str(round(squares2[4] * 100, 2)) + "%")
print("radius_worst: " + str(round(squares2[5] * 100, 2)) + "%")
print("texture_worst: " + str(round(squares2[6] * 100, 2)) + "%")

features = ['compactness_se', 'concavity_se', 'concave points_se', 'symmetry_se',
       'fractal_dimension_se', 'radius_worst', 'texture_worst',]
X = df5[features]


loadings = pca.components_.T * np.sqrt(pca.explained_variance_)
#fig = px.scatter(components_5, x=0, y=1)
total_var = pca.explained_variance_ratio_.sum() * 100

labels = {str(i): f"PC {i+1}" for i in range(n_components)}
labels['color'] = 'Median Price'

fig = px.scatter(
    components_u, x=0, y=1,
    title=f'Total Explained Variance: {total_var:.2f}%',
    labels={'0': 'PC 1', '1': 'PC 2', '2': 'PC 3'}
)
for i, feature in enumerate(features):
    fig.add_shape(
        type='line',
        x0=0, y0=0,
        x1=loadings[i, 0],
        y1=loadings[i, 1]
    )
    fig.add_annotation(
        x=loadings[i, 0],
        y=loadings[i, 1],
        ax=0, ay=0,
        xanchor="center",
        yanchor="bottom",
        text=feature,
    )


#fig = px.scatter(components_u, x=0, y=1, color_continuous_scale=px.colors.sequential.Blackbody)
fig.show()

Principle Components
PC1: 71.09%
PC2: 28.91%

Feature percentages in PC1
compactness_se: 0.0%
concavity_se: 0.0%
concave points_se: 0.0%
symmetry_se: 0.0%
fractal_dimension_se: 0.0%
radius_worst: 22.05%
texture_worst: 77.95%

Feature percentages in PC2
compactness_se: 0.0%
concavity_se: 0.0%
concave points_se: 0.0%
symmetry_se: 0.0%
fractal_dimension_se: 0.0%
radius_worst: 77.95%
texture_worst: 22.05%


First, we will subset our data to look at specific columns including the standard error of compactness, concavity, concave points, symmetry, and fractal dimension. As well as the 'worst' or largest(mean of the three largest values) of the radius and texture of the cancer cell nuclei. Subsetting our dataframe and performing PCA, we find that our first two components capture almost all of our subsetted data!

Seeing the feature percentages, we find that texture_worst captures 77.95% of variance on the first component and radius_worst  contributes about 22% of the rest on the first componenet. In terms of cell measurements, this makes sense as the largest measurements and most extreme will seemingly take over any account on standard error of other measurements.

Interestingly, the percentages are flipped on the second component. This must mean that out of these features, radius_worst and texture_worst contribute to mostly all of the variance and have a relationship togethor.

Looking at the loading vectors, we find that radius_worst and texture_worst have the greatest magnitude and thus explain the most amount of variance for each of the components, while all the other loadings are grouping around the same point because of their insignificance to the components.



In [ ]:
n_pcs = pca.components_.shape[0]
print("number of principal components: ", n_pcs)

important_components = [np.abs(pca.components_[i]).argmax() for i in range(n_pcs)]
print("indicies of important features: ", important_components)

important_features = [features[important_components[i]] for i in range(n_pcs)]
print("most important features ", important_features, "\n")

number of principal components:  2
indicies of important features:  [6, 5]
most important features  ['texture_worst', 'radius_worst'] 



Finding our important features once again from our components confirm our findings and our graph above that texture_worst and radius_worst are the important feautres

In [ ]:
important_feature_data = df5.loc[:, important_features]
#important_feature_data['Hue'] = X['Hue']



loadings = pca.components_.T * np.sqrt(pca.explained_variance_)
fig = px.scatter(important_feature_data,
                    x=important_features[0],
                    y=important_features[1],
                 trendline="ols", trendline_scope="overall")





fig.update_traces(marker_coloraxis=None)
fig.show()


Graghing texture_worst and radius_worst togethor on a scatter plot with a overall trend line, we see a positive correlation peaking through. This shows that there is indeed a distinct relationship between these two variables in general, which then carries over into the principal components.

#Breast Cancer data analysis 2

 Since this data is so cool, let us do some further PCA analysis on more subsetted data. This time on the worst or largest radius, texture, perimeter, area, and smoothness. Also the standard error on concave points, symmetry, and fractal dimension.

In [ ]:
pca = PCA(n_components=2)
X = df5.loc[:, ['concave points_se', 'symmetry_se',
       'fractal_dimension_se', 'radius_worst', 'texture_worst',
       'perimeter_worst', 'area_worst', 'smoothness_worst']]
components_b = pca.fit_transform(X)
#print(pca.explained_variance_ratio_)
#print(pca.feature_names_in_)

pca2 = PCA(svd_solver='auto', whiten=True)
pca2.fit(X)

#print(pca.components_)
#print(sum(pca.components_[0][0:6]))
#print(sum(pca.components_[1][0:6]))
#print(sum(pca.components_[0][0:6]) + sum(pca.components_[1][0:6]))
#print(sum(pca.explained_variance_ratio_))

#Print the PC percentages
print("Principle Components")
print("PC1: " + str(round(pca.explained_variance_ratio_[0] * 100, 2)) + "%")
print("PC2: " + str(round(pca.explained_variance_ratio_[1] * 100, 2)) + "%\n")


squares1 = [1, 2, 3, 4, 5, 6, 7, 8]
i = 0
for x in pca.components_[0]:
  squares1[i] = x * x
  i += 1

squares2 = [1, 2, 3, 4, 5, 6, 7, 8]
i = 0
for x in pca.components_[1]:
  squares2[i] = x * x
  i += 1

#print(sum(squares1))
#print(squares1)

#Print the feature percentages



print("Feature percentages in PC1")
print("concave points_se: " + str(round(squares1[0] * 100, 2)) + "%")
print("symmetry_se: " + str(round(squares1[1] * 100, 2)) + "%")
print("fractal_dimension_se: " + str(round(squares1[2] * 100, 2)) + "%")
print("radius_worst: " + str(round(squares1[3] * 100, 2)) + "%")
print("texture_worst: " + str(round(squares1[4] * 100, 2)) + "%")
print("perimeter_worst: " + str(round(squares1[5] * 100, 2)) + "%")
print("area_worst: " + str(round(squares1[6] * 100, 2)) + "%")
print("smoothness_worst: " + str(round(squares1[7] * 100, 2)) + "%\n")

print("Feature percentages in PC2")
print("concave points_se: " + str(round(squares2[0] * 100, 2)) + "%")
print("symmetry_se: " + str(round(squares2[1] * 100, 2)) + "%")
print("fractal_dimension_se: " + str(round(squares2[2] * 100, 2)) + "%")
print("radius_worst: " + str(round(squares2[3] * 100, 2)) + "%")
print("texture_worst: " + str(round(squares2[4] * 100, 2)) + "%")
print("perimeter_worst: " + str(round(squares2[5] * 100, 2)) + "%")
print("area_worst: " + str(round(squares2[6] * 100, 2)) + "%")
print("smoothness_worst: " + str(round(squares2[7] * 100, 2)) + "%")

features = ['concave points_se', 'symmetry_se',
       'fractal_dimension_se', 'radius_worst', 'texture_worst',
       'perimeter_worst', 'area_worst', 'smoothness_worst']
X = df5[features]


loadings = pca.components_.T * np.sqrt(pca.explained_variance_)
#fig = px.scatter(components_5, x=0, y=1)
total_var = pca.explained_variance_ratio_.sum() * 100

labels = {str(i): f"PC {i+1}" for i in range(n_components)}
labels['color'] = 'Median Price'

fig = px.scatter(
    components_b, x=0, y=1,
    title=f'Total Explained Variance: {total_var:.2f}%',
    labels={'0': 'PC 1', '1': 'PC 2', '2': 'PC 3'}
)

for i, feature in enumerate(features):
    fig.add_shape(
        type='line',
        x0=0, y0=0,
        x1=loadings[i, 0],
        y1=loadings[i, 1]
    )
    fig.add_annotation(
        x=loadings[i, 0],
        y=loadings[i, 1],
        ax=0, ay=0,
        xanchor="center",
        yanchor="bottom",
        text=feature,
    )


#fig = px.scatter(components_u, x=0, y=1, color_continuous_scale=px.colors.sequential.Blackbody)
fig.show()


NameError: ignored

Running PCA, we find our first principal component to capture 99.97% of our variance!!!! Awesome! This means all of our important data is essentially can be reduced to just one dimension.

On our first component, we find that area_worst captures 99.66% of our variance. And on the second component, we find that perimeter_worst captures 90.62% of variance.

This makes sense as perimeter and area are directly related to one another. So, in terms of cancer nucleic cells, the largest areas and perimeters seems to not only makeup the most important features, but correlate strongly as well.

The loadings of our components and their features tell us that perimeter_worst and area_worst have the greatest magnitude and describe the points closest to whatever direction the projected points of our data align with.

In [ ]:
n_pcs = pca.components_.shape[0]
print("number of principal components: ", n_pcs)

important_components = [np.abs(pca.components_[i]).argmax() for i in range(n_pcs)]
print("indicies of important features: ", important_components)

important_features = [features[important_components[i]] for i in range(n_pcs)]
print("most important features ", important_features, "\n")

AttributeError: ignored

Additionally, the important features of our data is the area_worst and perimeter_worst measurements of the cancer nucleic cells.

In [ ]:
important_feature_data = df5.loc[:, important_features]
#important_feature_data['Hue'] = X['Hue']



loadings = pca.components_.T * np.sqrt(pca.explained_variance_)
fig = px.scatter(important_feature_data,
                    x=important_features[0],
                    y=important_features[1],
                   )




fig.update_traces(marker_coloraxis=None)
fig.show()


Graphing our perimeter_worst and area_worst, we find that an extremely strong correlation with a positive slope.

#Pizza

Hello once more Boss. We also serve delicious pizza as part of our business on top of celluar cancer research.
(we try not to mix them up) Here, I used data from other pizza coporations to show what parts of pizza are the most correlated and important to craft the most delicious pizza.  

In [ ]:
dfa = pd.read_csv("/content/drive/MyDrive/Christian Kolker/Pizza.csv")
dfa = dfa.drop('brand', 1)
dfa = dfa.drop('id', 1)
#df5 = df5.drop('Unnamed: 32', 1)
dfa.head()


mois   prot    fat   ash  sodium  carb   cal
0  27.82  21.43  44.87  5.11    1.77  0.77  4.93
1  28.49  21.26  43.89  5.34    1.79  1.02  4.84
2  28.35  19.99  45.78  5.08    1.63  0.80  4.95
3  30.55  20.15  43.13  4.79    1.61  1.38  4.74
4  30.49  21.28  41.65  4.82    1.64  1.76  4.67

In [ ]:
scaler=StandardScaler()
df=scaler.fit_transform(dfa)
n_components = 2

pca = PCA(n_components=n_components)
components_a = pca.fit_transform(dfa)

total_var = pca.explained_variance_ratio_.sum() * 100

labels = {str(i): f"PC {i+1}" for i in range(n_components)}
labels['color'] = 'Median Price'

fig = px.scatter(
    components_a, x=0, y=1,
    title=f'Total Explained Variance: {total_var:.2f}%',
    labels={'0': 'PC 1', '1': 'PC 2', '2': 'PC 3'}
)
fig.show()
pca.explained_variance_ratio_

array([0.78010108, 0.18840661])

Here, we can see that in terms of our componenets, the majority of our important data is in two dimensions and we can see what the data is exactly for delicious pizzas.

In [ ]:
n_pcs = pca.components_.shape[0]
print("number of principal components: ", n_pcs)

important_components = [np.abs(pca.components_[i]).argmax() for i in range(n_pcs)]
print("indicies of important features: ", important_components)

important_features = [features[important_components[i]] for i in range(n_pcs)]
print("most important features ", important_features, "\n")

number of principal components:  2
indicies of important features:  [5, 0]
most important features  ['perimeter_worst', 'concave points_se'] 



Our analysis of the important features of PCA show that the amount of carbohydrates and moistness of a 100g sample of pizza are the most important parts.

In [ ]:
pca = PCA(n_components=2)

components_a1 = pca.fit_transform(dfa)
#print(pca.explained_variance_ratio_)
#print(pca.feature_names_in_)

pca2 = PCA(svd_solver='auto', whiten=True)
pca2.fit(X)

#print(pca.components_)
#print(sum(pca.components_[0][0:6]))
#print(sum(pca.components_[1][0:6]))
#print(sum(pca.components_[0][0:6]) + sum(pca.components_[1][0:6]))
#print(sum(pca.explained_variance_ratio_))

#Print the PC percentages
print("Principle Components")
print("PC1: " + str(round(pca.explained_variance_ratio_[0] * 100, 2)) + "%")
print("PC2: " + str(round(pca.explained_variance_ratio_[1] * 100, 2)) + "%\n")


squares1 = [1, 2, 3, 4, 5, 6, 7, 8]
i = 0
for x in pca.components_[0]:
  squares1[i] = x * x
  i += 1

squares2 = [1, 2, 3, 4, 5, 6, 7, 8]
i = 0
for x in pca.components_[1]:
  squares2[i] = x * x
  i += 1

#print(sum(squares1))
#print(squares1)

#Print the feature percentages
print("Feature percentages in PC1")
print("mois: " + str(round(squares1[0] * 100, 2)) + "%")
print("prot: " + str(round(squares1[1] * 100, 2)) + "%")
print("fat: " + str(round(squares1[2] * 100, 2)) + "%")
print("ash: " + str(round(squares1[3] * 100, 2)) + "%")
print("sodium: " + str(round(squares1[4] * 100, 2)) + "%")
print("carb: " + str(round(squares1[5] * 100, 2)) + "%")
print("cal: " + str(round(squares1[6] * 100, 2)) + "%\n")

print("Feature percentages in PC2")
print("mois: " + str(round(squares2[0] * 100, 2)) + "%")
print("prot: " + str(round(squares2[1] * 100, 2)) + "%")
print("fat: " + str(round(squares2[2] * 100, 2)) + "%")
print("ash: " + str(round(squares2[3] * 100, 2)) + "%")
print("sodium: " + str(round(squares2[4] * 100, 2)) + "%")
print("carb: " + str(round(squares2[5] * 100, 2)) + "%")
print("cal: " + str(round(squares2[6] * 100, 2)) + "%")

features = ['mois',	'prot',	'fat',	'ash',	'sodium',	'carb',	'cal']
X = dfa[features]


loadings = pca.components_.T * np.sqrt(pca.explained_variance_)
#fig = px.scatter(components_5, x=0, y=1)
total_var = pca.explained_variance_ratio_.sum() * 100

labels = {str(i): f"PC {i+1}" for i in range(n_components)}
labels['color'] = 'Median Price'

fig = px.scatter(
    components_a1, x=0, y=1,
    title=f'Total Explained Variance: {total_var:.2f}%',
    labels={'0': 'PC 1', '1': 'PC 2', '2': 'PC 3'}
)

for i, feature in enumerate(features):
    fig.add_shape(
        type='line',
        x0=0, y0=0,
        x1=loadings[i, 0],
        y1=loadings[i, 1]
    )
    fig.add_annotation(
        x=loadings[i, 0],
        y=loadings[i, 1],
        ax=0, ay=0,
        xanchor="center",
        yanchor="bottom",
        text=feature,
    )


#fig = px.scatter(components_u, x=0, y=1, color_continuous_scale=px.colors.sequential.Blackbody)
fig.show()


Principle Components
PC1: 78.01%
PC2: 18.84%

Feature percentages in PC1
mois: 7.67%
prot: 7.13%
fat: 7.78%
ash: 0.31%
sodium: 0.01%
carb: 77.1%
cal: 0.0%

Feature percentages in PC2
mois: 55.81%
prot: 0.31%
fat: 43.28%
ash: 0.16%
sodium: 0.06%
carb: 0.0%
cal: 0.38%


Looking at the statistics of each componenet, we find that the first component holding the most important parts of our pizza data take up about 77.1% of carbohydrates. This is also indicated by the large arrow that is labeled carbohydrates, meaning much more variance is on carbs in pizza. This means carbohydrates are the most meaningful parts of delicious pizza!

Additionally, moistness and the amount of fat of the pizza is a second important factor. This is shown in the second component that takes up 55.81% for moistness and 43.28% for fat. This makes sense as good pizza is determined by the fat, particularly by the cheese, and the moisture of the pizza!

Now, using data science we can now focus on the most important parts of pizza (carbohydrates, fatness, and moisture) to craft the perfect pizza!

#Conclusion


##PCA and findings

In conclusion, I have used PCA to find the most important data in our dataframes by reducing our dimensionality with our principal components and calculating the amount of variance each of the components explain. This variance tells us not only the important parts of our data, but it also allows us to easily interpret the data into something we can visualize and not bother with higher dimensions.

Additionally, the other parts of our data that did not contribute to the components prominently are considered noise and do not contribute much if not at all.

So, that is PCA and its awesome application.
